### DISCUSSION TITLE
int and uint changes

### DISCUSSION THREAD

**Discussion by Gavin-Williams** [amerkoleci/Vortice.Windows#470] [General]
I'm getting hundreds of errors due to some changes with methods asking for uint now and not being compatible with various code that interacts with them. For example...

IWICBitmapSourceSize.Width - integer

But PixelFormat.GetStride(Guid guid, uint width)

Was this change intentional? Usually, integer is selected for its broad compatibility, even though uint might make more sense as a container for valid values.

Or are we just expect to go through and type (uint) hundreds of times? (225 for me). And even making a few changes, It has ballooned out to 300. This is a nightmare.

What's the rationale in using int or uint in any situation?

**Comment by amerkoleci** [comment]
Hi,
Sorry for causing trouble to you, the idea was to keep same sign as native types, would you be able to fix it and use uint or would you prefer reverting back to int?

Thanks,
Amer

**Reply by Gavin-Williams** [reply]
Look, I have to admit, I've struggled to decide which is best myself. I asked about the use of uint in the Dx channel. TBH I wasn't that impressed with the response - I mean, what they said is fine, but it's just so subjective.

However. Making it compatible with Dx I suppose is fine. I'd like to hear from others who have much larger code bases than mine or who are more experiences. You are certainly a better programmer than me, your grasp of native is much stronger. So, I will follower your lead.

But there may be some things we can do to make it go smoother. For example, make sure that all the interactive types use uint as well. For example...

IWICBitmapSource uses uint in Dx, but Vortice returns a SizeI (integer type). If those cases are covered, maybe the error count could be reduce.

There's probably no getting around the mess that is created by the change. I just have to go through and decide case by case what to do.

**Comment by MarcioAB** [comment]
My Vortice program and my coding experienceis are both not large but I did the conversion in 2 hours work. I was using -1 as significative value so I had to change strategy.

**Comment by h1cks** [comment]
I think that there are a couple of classes where INT is the value based upon the fact there are a couple of c# conversion issues. These are based upon how c# seems to enforce type inference.

For example in Texture2DShaderResourceView there is a Miplevel which is UINT, but you can set this to -1 (now, this is sort of a hack from what i recall for the c++ handling), but in c# you really need to set uint.MaxValue instead to represent the same byte format.

**Comment by h1cks** [comment]
BTW.  After I moved from INT to UINT, I seem to have improved my framerate by around 10 to 20 percent.  Anecdotal atm.  My project is very large btw.

Notes of pain:

Using Size, Box etc are all int based, so there are a few conversions.  On the whole though, not so bad.
